src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs

```diff
@@ -67,13 +68,17 @@ public void Authenticate(IConnection connection, ConnectionDescription descripti
 
             using (var conversation = new SaslConversation(description.ConnectionId))
             {
-                var currentStep = _mechanism.Initialize( connection, conversation, description);
-                var command = CreateStartCommand(currentStep);
-
+                ISaslStep currentStep;
+                BsonDocument command;
                 var speculativeAuthenticateResult = description.IsMasterResult.SpeculativeAuthenticate;
-                if (speculativeAuthenticateResult != null)
+                if (_speculativeFirstStep != null && speculativeAuthenticateResult != null)
```

Comment:
is there a chance that only one of these fields will be filled?

Reply:
Hey @DmitryLukyanov ,
Thanks for the insightful question :) It took me a while to re-derive the answer!
So, we can have two cases where only one of the fields is not null.
The first case is simple:

`speculativeFirstStep != null && speculativeAuthenticateResult == null` : We tried to speculatively authenticate, but the server doesn't support this feature.

The second case is a bit more complicated:

`speculativeFirstStep == null && speculativeAuthenticateResult !=null`: We have at multiple credentials with no auth source (e.g. defaultAuthenticators). The first `DefaultAuthenticator` successfully speculatively authenticates with ScramSha256, and merges the `isMasterResult`: https://github.com/vincentkam/mongo-csharp-driver/blob/0944b1f474b0711d03675bbd9c21879e48bb2ec6/src/MongoDB.Driver.Core/Core/Authentication/DefaultAuthenticator.cs#L102
The second `DefaultAuthenticator` will not attempt to speculatively auth because the new isMasterResult will have `SaslSupportedMechs` (from the first default authenticator's merge). Therefore, a new  `ScramSha256Authenticator` is created. This `ScramSha256Authenticator` will see `speculativeAuthenticateResult != null but its `speculativeFirstStep` will be null.

You can test this out if you'd like by running this test: https://github.com/mongodb/mongo-csharp-driver/blob/3b807e658c1459682ebecd12bf3e4e57884ef741/tests/MongoDB.Driver.Tests/AuthenticationTests.cs#L169
and commenting out the null check on `speculativeFirstStep`.

This case is not covered by the spec because we're on of the few (only?) drivers that still supports multiple credentials. According to @jyemin , even Java dropped support a while ago :)

Reply:
thanks for clarification